### PR TITLE
Fixed `App.current_window` on macOS, for when a dialog is in focus

### PR DESCRIPTION
**Pull request by proneon267**
* Previously on macOS, when a dialog was in focus, `App._impl.current_window` returned the `NSPanel` object instead of the `NSWindow` object, causing an AttributeError. 
Error:
```zsh
  File "/Users/proneon267/toga/testbed/tests/app/test_desktop.py", line 390, in test_current_window
    assert app.current_window == window2
  File "/Users/proneon267/toga/core/src/toga/app.py", line 834, in current_window
    return window.interface
           ^^^^^^^^^^^^^^^^
  File "/Users/proneon267/venv/lib/python3.12/site-packages/rubicon/objc/api.py", line 1097, in __getattr__
    raise AttributeError(
AttributeError: rubicon.objc.api.ObjCInstance _NSAlertPanel has no attribute interface
```
This PR corrects this behaviour and now `App.current_window` returns the host window of the dialog. This bug was identified while working on #2473 

## PR Checklist:

- [x] All new features have been tested
- [x] All new features have been documented
- [x] I have read the **CONTRIBUTING.md** file
- [x] I will abide by the code of conduct

### CODE REVIEW

**Comment by proneon267** [comment]
A bit of an off-topic, but is this comment still valid:
https://github.com/beeware/toga/blob/75e1fd97a52bfc408793adbed261c9e8be54021e/testbed/tests/app/test_desktop.py#L355-L358
I thought I had corrected this behaviour and that GTK also returned None, when all windows are hidden:
https://github.com/beeware/toga/blob/75e1fd97a52bfc408793adbed261c9e8be54021e/gtk/src/toga_gtk/app.py#L237-L239

File: testbed/tests/app/test_desktop.py [hunk 1]
```diff
@@ -370,25 +368,45 @@ async def test_current_window(app, app_probe, main_window):
...
 
     app.current_window = window3
-    await window1_probe.wait_for_window("Window 3 is current")
+    window3_dialog_task = app.loop.create_task(window3.dialog(window3_info_dialog))
+    await main_window_probe.wait_for_window("Window 3 is current")
     if app_probe.supports_current_window_assignment:
         assert app.current_window == window3
+    await app_probe.redraw("select 'OK'")
+    # Cancel the task to avoid dangling
+    window3_dialog_task.cancel()
```
Comment (freakboy3742) [inline testbed/tests/app/test_desktop.py:409]: I don't think we need to do the "in the presence of a dialog" test multiple times. We should still be testing the "normal" case of "just assigning a new window"; we only need to do one additional check of "...and if there's a current dialog, that doesn't get in the way".

We *should*, however, confirm that when a dialog is dismissed, but the current window *hasn't* been modified that the current window is still what it should be.

File: testbed/tests/app/test_desktop.py [hunk 2]
```diff
@@ -370,25 +368,45 @@ async def test_current_window(app, app_probe, main_window):
...
 
     app.current_window = window2
-    await window1_probe.wait_for_window("Window 2 is current")
+    window2_dialog_task = app.loop.create_task(window2.dialog(window2_info_dialog))
+    await main_window_probe.wait_for_window("Window 2 is current")
     if app_probe.supports_current_window_assignment:
         assert app.current_window == window2
+    await app_probe.redraw("select 'OK'")
+    # Cancel the task to avoid dangling
+    window2_dialog_task.cancel()
```
Comment (freakboy3742) [inline testbed/tests/app/test_desktop.py:397]: I'm unclear why this cancel is needed. What sort of "dangling" can occur? Why is this style of cancel needed here when it isn't needed on any other dialog test?

**Review by freakboy3742 — CHANGES_REQUESTED** [review]
Broadly makes sense; a couple of questions about the testing strategy inline.

File: testbed/tests/app/test_desktop.py (hunk 1 above)
Comment (proneon267) [reply testbed/tests/app/test_desktop.py:409]: I agree, I have made the "in presence of dialog test" to be tested only on window1.

File: testbed/tests/app/test_desktop.py (hunk 2 above)
Comment (proneon267) [reply testbed/tests/app/test_desktop.py:397]: I was getting error for the app still having dangling tasks when exiting the test. On deeper inspection, the error was probably telling me that the dialog is not dismissed, even after calling `await app_probe.redraw("select 'OK'")`

**Comment by proneon267** [comment]
It seems like there are bugs in setting up dialogs.
## When the Dialog setup uses `main_window_probe`:
The dialog automatically gets dismissed as soon as it is displayed, when the dialog setup uses `main_window_probe`:
https://github.com/beeware/toga/blob/c667c3e571f51366b76a6f3035d6d7517f17811e/testbed/tests/app/test_desktop.py#L394-L395

### Test Run 1:https://github.com/beeware/toga/pull/2926/commits/c667c3e571f51366b76a6f3035d6d7517f17811e
|For dialog setup|For showing the dialog|Works?|
|---|---|---|
|`window1_probe.setup_info_dialog_result(info_dialog) `|`await window1.dialog(info_dialog) `|❌|

https://github.com/beeware/toga/blob/c667c3e571f51366b76a6f3035d6d7517f17811e/testbed/tests/app/test_desktop.py#L387-L409

### Test Result:
```zsh
FAILED tests/app/test_desktop.py::test_current_window - AssertionError: The dialog is not in focus
assert '_NSAlertPanel' in '<TogaWindow: 0x7fe515f4de50>'
 +  where '<TogaWindow: 0x7fe515f4de50>' = str(<ObjCInstance: TogaWindow at 0x10eb7ab30: <TogaWindow: 0x7fe515f4de50>>)
 +    where <ObjCInstance: TogaWindow at 0x10eb7ab30: <TogaWindow: 0x7fe515f4de50>> = <ObjCInstance: NSApplication at 0x10df9d640: <NSApplication: 0x7fe513748e70>>.keyWindow
 +      where <ObjCInstance: NSApplication at 0x10df9d640: <NSApplication: 0x7fe513748e70>> = <toga_cocoa.app.App object at 0x10e2daa90>.native
 +        where <toga_cocoa.app.App object at 0x10e2daa90> = <testbed.app.Testbed object at 0x10db7e770>._impl
 ```
The test fails as the dialog gets dismissed as soon as it is displayed. Hence, it fails the dialog focus test.

## When the Dialog setup uses `app_probe`:
The dialog remains open and the tests do not continue

### Test Run 2:https://github.com/beeware/toga/pull/2926/commits/75f6e41d33e057426cddfc3c4d36a861b2815d21
|For dialog setup|For showing the dialog|Works?|
|---|---|---|
|`app_probe.setup_info_dialog_result(info_dialog) `|`await window1.dialog(info_dialog) `|❌|

https://github.com/beeware/toga/blob/75f6e41d33e057426cddfc3c4d36a861b2815d21/testbed/tests/app/test_desktop.py#L387-L409

### Test Result:
The dialog is shown, but the tests do not continue and the app gets into a deadlock state(i.e, the app doesn't respond while the dialog is showing). Also, the dialog cannot be dismissed.

## When the Dialog setup uses `app_probe` and `create_task()`:
The dialog is shown and the tests continue.

### Test Run 3:https://github.com/beeware/toga/pull/2926/commits/1e04814c16867ce1b99f1aea680af5bfe799781c
|For dialog setup|For showing the dialog|Works?|
|---|---|---|
|`app_probe.setup_info_dialog_result(info_dialog) `|` app.loop.create_task(window1.dialog(info_dialog)) `|✅|

https://github.com/beeware/toga/blob/1e04814c16867ce1b99f1aea680af5bfe799781c/testbed/tests/app/test_desktop.py#L390-L402

### Test Result:
The test passes and the dialog is in focus on the native backend.

### Problem with this approach:
But, the dialog still cannot be dismissed and remains hanging. So, it will complain about the app having dangling task when exiting the test, hence we have to use `dialog_task.cancel()`

File: testbed/tests/app/test_desktop.py [hunk 3]
```diff
@@ -370,26 +368,47 @@ async def test_current_window(app, app_probe, main_window):
...
+    dialog_task = app.loop.create_task(window1.dialog(info_dialog))
+    await main_window_probe.wait_for_window("Displayed window1 modal info dialog")
+    # The public API should report that current window is the specified window
+    if app_probe.supports_current_window_assignment:
+        assert app.current_window == window1
+    # But, the backend should be reporting that the current window is the dialog
+    if toga.platform.current_platform == "macOS":
+        assert "_NSAlertPanel" in str(
+            app._impl.native.keyWindow
+        ), "The dialog is not in focus"
```
Comment (proneon267) [inline testbed/tests/app/test_desktop.py:405]: This test is only to test that the test is actually testing the situation of "in presence of dialog" and not testing the situation where the dialog immediately exits.

File: testbed/tests/app/test_desktop.py [hunk 4]
```diff
@@ -370,26 +368,47 @@ async def test_current_window(app, app_probe, main_window):
...
 
+    # When a dialog is in focus, `app.current_window` should
+    # return the window from which the dialog was initiated.
+    info_dialog = toga.InfoDialog("Info", "Some info")
+    app_probe.setup_info_dialog_result(info_dialog)
+
+    app.current_window = window1
+    await main_window_probe.wait_for_window("Window 1 is current")
+    dialog_task = app.loop.create_task(window1.dialog(info_dialog))
+    await main_window_probe.wait_for_window("Displayed window1 modal info dialog")
```
Comment (freakboy3742) [inline testbed/tests/app/test_desktop.py:397]: I'm not convinced this is a good test. 

Firstly, the assignment to window1 is redundant - there's nothing special about window 1, so that's a step that isn't required.

But, more importantly, this entirely depends on the exact sequencing of events. It's using `create_task()` to schedule a coroutine on the event loop - and then *hoping* that all the checks about window assignment are performed before that task completes. 

In practice, that seems highly likely... but it's not guaranteed. And it also leads to the "dangling task" issue, as there's no guarantee that the task will have ended by the time the test is cleaning up.

A better approach would be to modify the `setup_info_dialog_result` probe to accept an optional method that contains "pre-close" test logic, and invoke that method as required before closing.

**Review by freakboy3742 — CHANGES_REQUESTED** [review]
The fix looks good; the issues are entirely related to testing.

File: testbed/tests/app/test_desktop.py (hunk 4 above)
Comment (proneon267) [reply testbed/tests/app/test_desktop.py]: Apologies for the delayed response; I got caught up with a few things. Anyway, Thank you for suggesting the test strategy.

**Comment by proneon267** [comment]
On macOS, when a dialog is dismissed, the window from which the dialog was initiated becomes the current_window, regardless of whether the current_window is modified. But, this is not the behavior on other backends(gtk, winforms). Should I modify this behavior?(If yes, on which backend?), or should I just document this as a native platform dependent behavior?

**Comment by freakboy3742** [comment]
> On macOS, when a dialog is dismissed, the window from which the dialog was initiated becomes the current_window, regardless of whether the current_window is modified.
> 
> But, this is not the behavior on other backends(gtk, winforms). Should I modify this behavior?(If yes, on which backend?), or should I just document this as a native platform dependent behavior?

I'd argue that's a platform quirk. We shouldn't be overriding default platform behavior unless there's a *very* good reason to do so.

File: android/tests_backend/app.py [hunk 5]
```diff
@@ -42,6 +42,9 @@ def logs_path(self):
     def assert_app_icon(self, icon):
         pytest.xfail("Android apps don't have app icons at runtime")
 
+    def assert_dialog_in_focus(self, dialog):
```
Comment (freakboy3742) [inline android/tests_backend/app.py:45]: This (and other "empty" implementations) should either document *why* they're empty implementations, or provide a token "dialog is visible" test.

File: android/tests_backend/dialogs.py [hunk 6]
```diff
@@ -28,7 +28,7 @@ async def _close_dialog():
 
         dialog._impl.show = automated_show
 
-    def setup_info_dialog_result(self, dialog):
+    def setup_info_dialog_result(self, dialog, pre_close_test_method=None):
```
Comment (freakboy3742) [inline android/tests_backend/dialogs.py:31]: This is a foot-gun waiting to go off. If this extension point is being added, then it needs to be added (and added *everywhere*), not just stubbed out enough to make the interpreter stop complaining. Otherwise, someone will try to use the method, assume the test is executed, and then be surprised when it isn't.

File: core/src/toga/app.py [hunk 7]
```diff
@@ -827,7 +827,12 @@ def show_cursor(self) -> None:
 
     @property
     def current_window(self) -> Window | None:
-        """Return the currently active window."""
+        """The window currently in active focus.
+
+        On macOS, when a :any:`Dialog` is dismissed, the window from which the dialog
+        was initiated becomes the :any:`App.current_window`, regardless of whether the
+        :any:`App.current_window` property was modified.
```
Comment (freakboy3742) [inline core/src/toga/app.py:834]: I'm not convinced this needs to be documented. Yes, it's a platform quirk; but "macOS behaves like macOS" isn't news. If we want to get technical, I'm going to bet that GTK has some pretty quirky behavior here too, because focus management is a window manager concern. The WM we're using for test purposes returns focus to the old current window - but a "focus follows mouse" WM almost certainly won't.

**Review by freakboy3742 — CHANGES_REQUESTED** [review]
The mechanics of the test you've added all look good. However, you've gone down the "make the interpreter be quiet" path, rather than finishing the job. If we're adding a feature, *we add the feature*. We don't partially add the feature in a way that is a bug in waiting for future users who assume that the API behaves the same everywhere.

File: core/src/toga/app.py (hunk 7 above)
Comment (proneon267) [reply core/src/toga/app.py]: I agree with you, so I've removed the note.

**Comment by proneon267** [comment]
I will implement the probe features across all the required backends. But it would take some time. I know you have told me to work on one PR at one time, but in the meantime, can you please review #2473, when you are free? This PR isn't related to #2473, so #2473 can be reviewed independently of this PR.

File: android/src/toga_android/dialogs.py [hunk 8]
```diff
@@ -22,7 +22,7 @@ def show(self, host_window, future):
 
         if self.native:
             # Show the dialog. Don't differentiate between app and window modal dialogs.
-            self.native.show()
+            self.native_alert_dialog = self.native.show()
```
Comment (freakboy3742) [inline android/src/toga_android/dialogs.py:25]: What is the purpose of this assignment? Firstly, `self.native` should be a reference to the native object; secondly [`Dialog.show()` returns void](https://developer.android.com/reference/android/app/Dialog#show()), so this value should always be None`.

File: android/tests_backend/app.py [hunk 9]
```diff
@@ -42,6 +42,11 @@ def logs_path(self):
...
+        assert (
+            dialog._impl.native_alert_dialog.isShowing() is True
```
Comment (freakboy3742) [inline android/tests_backend/app.py:47]: Related to above - why isn't this just `dialog._impl.native.isShowing()`?

File: cocoa/tests_backend/app.py [hunk 10]
```diff
@@ -190,6 +190,14 @@ def activate_menu_close_all_windows(self):
     def activate_menu_minimize(self):
         self._activate_menu_item(["Window", "Minimize"])
 
+    def assert_dialog_in_focus(self, dialog):
+        # Cannot directly compare `dialog._impl.native`(`NSAlert`) and
+        # `self.app._impl.native.keyWindow`(`_NSAlertPanel`) objects.
+        # Hence, do a string comparison instead.
+        assert str(dialog._impl.native.objc_class.__name__) in str(
+            self.app._impl.native.keyWindow.objc_class.__name__
+        ), "The dialog is not in focus"
```
Comment (freakboy3742) [inline cocoa/tests_backend/app.py:199]: No issue with a string comparison if it's needed... but doesn't `NSAlert.window` give you the window for the alert (which should be the keyWindow when the dialog is showing, AFAICT)?

File: gtk/src/toga_gtk/app.py [hunk 11]
```diff
@@ -236,7 +236,18 @@ def show_cursor(self):
...
+            # If the focus is on a dialog, then return its host window
+            for window in self.interface.windows:
+                dialog_impl = getattr(window._impl, "dialog_impl", None)
+                if dialog_impl:
+                    if dialog_impl.native.is_visible():  # pragma: no branch
+                        # Marking as no branch, since a window having dialog_impl
+                        # will have its dialog visible, as all dialogs are shown
+                        # in modal style.
+                        return window._impl
+            return current_window
```
Comment (freakboy3742) [inline gtk/src/toga_gtk/app.py:249]: I'm not clear what's going on with this change - is this something that was revealed as a result of the extra testing? It seems to be trying to protecting against a change of current window while a dialog is visible... but it's not clear to me how that would be possible.

File: winforms/src/toga_winforms/app.py [hunk 12]
```diff
@@ -251,6 +251,20 @@ def get_current_window(self):
         for window in self.interface.windows:
             if WinForms.Form.ActiveForm == window._impl.native:
                 return window._impl
+
+            # If the focus is on a dialog, then return its host window
```
Comment (freakboy3742) [inline winforms/src/toga_winforms/app.py:255]: As with the GTK backend, it's not clear why this change is required.

**Review by freakboy3742 — CHANGES_REQUESTED** [review]
So - this seems to have introduced a bunch of new behavior into the GTK and Winforms backend, and I'm not entirely clear why. This isn't just "adding tests to verify that the dialog is visible", which was the state of the PR last time I reviewed it - it seems to have dramatically changed what "current window" means on those platforms.

Was the old implementation incorrect? Or is this new behavior that then also required tests? It seems like the latter, because the new behavior seems to be aimed at catching cases where current window being re-assigned *while* a dialog is visible. 

Even then, it's not clear to me why this is required. If a dialog is window-modal, that doesn't prevent you from giving another window focus; if a dialog is app-modal, you can't change the current window, because only the dialog can have focus.

So - what's the motivation here?

**Comment by proneon267** [comment]
I agree I should have discussed the behavior of `App.current_window` before changing it. 
On testing I had found:
* On macOS, when a dialog was in focus, `App._impl.current_window` returned the `NSPanel` object instead of the `NSWindow` object, causing an AttributeError. 
Error:
```zsh
  File "/Users/proneon267/toga/testbed/tests/app/test_desktop.py", line 390, in test_current_window
    assert app.current_window == window2
  File "/Users/proneon267/toga/core/src/toga/app.py", line 834, in current_window
    return window.interface
           ^^^^^^^^^^^^^^^^
  File "/Users/proneon267/venv/lib/python3.12/site-packages/rubicon/objc/api.py", line 1097, in __getattr__
    raise AttributeError(
AttributeError: rubicon.objc.api.ObjCInstance _NSAlertPanel has no attribute interface
```
* On WinForms, when a dialog was in focus, `App._impl.current_window` returned `None`, instead of the window from which the dialog originated.

* On Gtk, when a dialog was in focus, `App._impl.current_window` returned "the most recent window that had active focus and is still currently visible", instead of the window from which the dialog originated.

Currently, I have changed the behavior of `App.current_window` to return the window from which the dialog originated 

But, what is actually the desired behavior and what should `App.current_window` return when a dialog is in focus?

**Comment by freakboy3742** [comment]
> Currently, I have changed the behavior of `App.current_window` to return the window from which the dialog originated
> 
> But, what is actually the desired behavior and what should `App.current_window` return when a dialog is in focus?

I can't say I have any strong opinions, other than "it definitely should not raise an error" and "the implementation shouldn't be exceptionally complex". 

Returning the window that the dialog is modal to would make sense - but that then leaves the question of what happens to app-modal windows. But if the price for that is an exceptionally complex implementation that involves iterating over every window... that API becomes a *lot* less desirable.

File: android/src/toga_android/dialogs.py (hunk 8 above)
Comment (proneon267) [reply android/src/toga_android/dialogs.py:25]: Actually `self.native` is a `AlertDialog.Builder` not a `AlertDialog`
https://github.com/beeware/toga/blob/3376c069cca85fb3c05523d3134f6842ed13f650/android/src/toga_android/dialogs.py#L32-L46
```
self.native = AlertDialog.Builder(toga.App.app.current_window._impl.app.native)
```
returns an `AlertDialog.Builder`: https://developer.android.com/reference/android/app/AlertDialog.Builder#Builder(android.content.Context)

So, doing `AlertDialog.Builder.show()`:  https://developer.android.com/reference/android/app/AlertDialog.Builder#show()
```
self.native_alert_dialog = self.native.show()
```
returns an `AlertDialog`


And `AlertDialog` has the `isShowing()` method: https://developer.android.com/reference/android/app/Dialog#isShowing()
which is used in the probe.

File: android/tests_backend/app.py (hunk 9 above)
Comment (proneon267) [reply android/tests_backend/app.py:47]: Same reason as above: https://github.com/beeware/toga/pull/2926/files#r1834096789

`AlertDialog` has the `isShowing()` method: https://developer.android.com/reference/android/app/Dialog#isShowing()
which is used in the probe.

**Comment by proneon267** [comment]
>I can't say I have any strong opinions, other than "it definitely should not raise an error" and "the implementation shouldn't be exceptionally complex".

>Returning the window that the dialog is modal to would make sense - but that then leaves the question of what happens to app-modal windows. But if the price for that is an exceptionally complex implementation that involves iterating over every window... that API becomes a lot less desirable.

I think the macOS implementation is simple, but only the winforms and gtk implementations are complex. So, for winforms and gtk, we could be to have a list of currently showing dialogs with each dialog tracking whether it has the active focus. Then in `App._impl.get_current_window()`, we could iterate through this dialogs_list and return the host window of the dialog having active focus. As for app modal dialog, we could return either the `main_window` or `None`.

However, the winforms implementation might still be a bit complex, as native dialogs like `MessageBox` do not expose any native object, which we can use to check for focus or visibility. Instead all we can get is a hwnd, which with some trickery can be used to check if it is in active focus.

Since, I have implemented one form of complex implementation, it's only fair that I implement the other :) , before we take the final decision.

File: cocoa/tests_backend/app.py (hunk 10 above)
Comment (proneon267) [reply cocoa/tests_backend/app.py:196]: You are right. Thank you :)

**Comment by proneon267** [comment]
Currently, the behavior for:
* For window modal dialogs - `App.current_window` will return the window from which the dialog originated
* For app modal dialogs - `App.current_window` will return `None`, since logically the dialog shouldn't belong to any window and also because no host_window is passed to app modal dialogs:
https://github.com/beeware/toga/blob/3376c069cca85fb3c05523d3134f6842ed13f650/core/src/toga/app.py#L794-L800

**Review by freakboy3742 — CHANGES_REQUESTED** [review]
A few comments inline about some minor details; but more importantly, I'm not convinced the "current window is None when an app-modal dialog is shown" interpretation makes sense. 

Toga doesn't consider a dialog to be a window; and a window can be current when a dialog is shown. I'd argue it doesn't matter what type of dialog has been displayed - whatever window *was* the current window should still be the current window. That might be nothing (if there are no open windows); or it might be an actual window.

And, as far as I can tell, this is also the easiest implementation - this is what falls out by default from both GTK and Winforms (removing the need for all the extra dialog tracking on those platforms; and on macOS, adding the "pre-display current window" as something that is tracked on app-modal dialogs seems like it should be reasonably straightforward (and something that can be stored on the *dialog*)

The other detail worth noting is that if we've identified there's a clear difference of behavior between app-modal and window-modal dialogs... that is something that needs to be tested.

**Comment by proneon267** [comment]
The iOS test failure is unrelated to this PR.

File: cocoa/src/toga_cocoa/app.py [hunk 13]
```diff
@@ -109,6 +109,7 @@ def __init__(self, interface):
 
         # Create the lookup table for commands and menu items
         self._menu_items = {}
+        self._active_window_before_dialog = None
```
Comment (freakboy3742) [inline cocoa/src/toga_cocoa/app.py:112]: Quoting from my previous review: 

> ... on macOS, adding the "pre-display current window" as something that is tracked on app-modal dialogs seems like it should be reasonably straightforward **(and something that can be stored on the dialog)** 

(emphasis added)

Storing the state on the app means there's a circular dependency on setting the variable as soon as you have a one window with a dialog.

File: gtk/tests_backend/dialogs.py [hunk 14]
```diff
@@ -21,23 +22,42 @@ def _default_close_handler(self, dialog, gtk_result):
...
+                        if close_handler:
+                            close_handler(dialog, gtk_result)
+                        else:
+                            self._default_close_handler(dialog, gtk_result)
+                    except Exception as e:
+                        # An error occurred closing the dialog; that means the dialog
+                        # isn't what as expected, so record that in the future.
+                        future.set_exception(e)
+
+            asyncio.create_task(_close_dialog(), name="close-dialog")
```
Comment (freakboy3742) [inline gtk/tests_backend/dialogs.py:53]: Why has this become an async task?

File: testbed/tests/app/test_desktop.py [hunk 15]
```diff
@@ -348,50 +348,84 @@ async def test_show_hide_cursor(app, app_probe):
...
+
+        # But, the backend should be reporting that the dialog is in focus
+        app_probe.assert_dialog_in_focus(dialog)
+
+    # Test in presence of window modal dialog
+    previous_active_window = app.current_window
+    window_modal_info_dialog = toga.InfoDialog("Window Modal Dialog Info", "Some info")
+    main_window_probe.setup_info_dialog_result(
+        window_modal_info_dialog,
+        pre_close_test_method=partial(
```
Comment (freakboy3742) [inline testbed/tests/app/test_desktop.py:406]: Given that the dialog is available at the point the test method is being invoked (and would be of general utility to *any* dialog test method), and the previous_active_window is a constant - why not make `dialog` a required first argument of the test method. and make the usage of `previous_active_window` a closure? (i.e., the test method referencing a variable that is outside the context of the method).

File: winforms/src/toga_winforms/app.py [hunk 16]
```diff
@@ -249,10 +251,17 @@ def show_cursor(self):
     ######################################################################
 
     def get_current_window(self):
-        for window in self.interface.windows:
-            if WinForms.Form.ActiveForm == window._impl.native:
-                return window._impl
-        return None
+        if self._active_window_before_dialog:
```
Comment (freakboy3742) [inline winforms/src/toga_winforms/app.py:254]: Is this even needed? Do dialogs appear as ActiveForm?

**Review by freakboy3742 — CHANGES_REQUESTED** [review]
A couple of issues flagged inline - mostly associated with the request from the previous review to keep the stateful data on the dialog.

File: gtk/tests_backend/dialogs.py (hunk 14 above)
Comment (proneon267) [reply gtk/tests_backend/dialogs.py:53]: I had to add a slight delay to allow the dialog to show up:
https://github.com/beeware/toga/blob/f4d6c64766169cbc292701f9701997eb07b9063b/gtk/tests_backend/dialogs.py#L35-L53
So, I had made it an async task.

File: cocoa/src/toga_cocoa/app.py (hunk 13 above)
Comment (proneon267) [reply cocoa/src/toga_cocoa/app.py:112]: I think I didn't understand what you meant. Did you meant to store the state on the native dialog object(i.e, on the NSAlert object itself?) or on the `dialog._impl`? When we get the key window in presence of a dialog with `NSApplication.keyWindow`, it returns the associated private `NSWindow` instance(i.e., `_NSAlertPanel`). So, how do we get to the `_impl` or to the `NSAlert` object from a `_NSAlertPanel`?

Also, I now realize that storing the previous active window may not be the correct thing to do here. For example:
* There are 2 windows open(`Window A`, `Window B`) and `Window A` has the current focus. Now, an `app modal dialog` opens and it stores `Window A` as the previously active window. 
* Since, the dialog is app modal, the user can manually change focus to other window, so the user manually changes focus to `window B` and then again changes focus on to the `app modal dialog`. 
* Now, if the app were to query the current window then it would return the cached `Window A` as the previously active window, when in reality `Window C` was the previously active window. Hence, giving incorrect results.

This is not a problem on winforms, as only 1 dialog can be shown at any given time. So, focus can't be changed to other windows of the app(Until the dialog is not dismissed). But on winforms, there is no way of knowing whether the focus is on a dialog or on something outside the app (`Form.ActiveForm` returns `None` in both cases). And even if the focus cannot be changed to other windows of the app(while the dialog is not dismissed), but the focus can be changed to other apps. So, returning the cached previously active window would be incorrect.

The more I look at the problem and the solution that this PR implements, the more I realize that I am going against the native behavior of the platform. I am trying to bend something that is not meant to be bent in the way I want. And the result is fragile and complicated.

Maybe the correct solution is to just prevent the error on macOS, leave the other platforms with their native behavior and just document the native platform quirks.

Comment (freakboy3742) [reply cocoa/src/toga_cocoa/app.py:112]: > I think I didn't understand what you meant. Did you meant to store the state on the native dialog object(i.e, on the NSAlert object itself?) or on the `dialog._impl`? 

Either - just *not* the app.
 
> Also, I now realize that storing the previous active window may not be the correct thing to do here. For example:

Yes - this is exactly why I was flagging the state variable on the app as a problem - dialogs are separate from the app, and there's complex interactions at play here.

> The more I look at the problem and the solution that this PR implements, the more I realize that I am going against the native behavior of the platform. I am trying to bend something that is not meant to be bent in the way I want. And the result is fragile and complicated.
> 
> Maybe the correct solution is to just prevent the error on macOS, leave the other platforms with their native behavior and just document the native platform quirks.

... which is what I was suggesting before we went down this rabbit hole :-) The most recent iterations here are adding a lot of complex logic to handle a specific edge case that doesn't really have a clear and obvious answer. If a dialog is showing, the "current window" *is* a vague concept. 

Not crashing is the primary requirement. Beyond that, I don't see any problem with returning as much of an answer as any given platform will allow, which means that there may not be a current window on some platforms when a dialog is showing - and then documenting that as a platform quirk.

**Comment by proneon267** [comment]
I have removed any behavioral changes from platforms other than macOS. But, I have kept the "pre-close test" mechanisms, and modified the test in order to make sure `App.current_window` doesn't raise any exceptions in presence of the dialog.

File: testbed/tests/app/test_desktop.py [hunk 17]
```diff
@@ -348,50 +348,72 @@ async def test_show_hide_cursor(app, app_probe):
...
         assert app.current_window == window3
 
+    # When a dialog is in focus, app.current_window should return the
+    # previously active window.
+    def test_current_window_in_presence_of_dialog(dialog):
+        app_probe.assert_dialog_in_focus(dialog)
+
+        # Accessing current_window in presence of dialog shouldn't raise any exceptions.
+        _ = app.current_window
+
```
Comment (proneon267) [inline testbed/tests/app/test_desktop.py:398]: Should I wrap this around a try...except...pytest.fail() block? I have found that just invoking the property without the fix, makes the test to fail automatically. So, I am not explicitly failing the test. Is this correct?

**Comment by proneon267** [comment]
Also, I have been wanting to cut down on the number of commits in my PR. Would there be any problem if I squash all of my previous commits in this PR into a single commit?

File: testbed/tests/app/test_desktop.py (hunk 17 above)
Comment (freakboy3742) [reply testbed/tests/app/test_desktop.py:398]: An explicit fail isn't needed. The underlying problem we're is that accessing the attribute doesn't raise an error; any error under test conditions will be reported, so there's nothing gained by turning an AttributeError into an explicit test failure.

**Review by freakboy3742 — APPROVED** [review]
Looks like we've got there! Thanks for the fix!

As for collapsing the commits - I can (and was planning to) do that as part of the merge of the PR.
